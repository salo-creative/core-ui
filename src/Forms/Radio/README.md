# Radio

Radio takes an 'options' prop and uses that to generate a number of CheckBox elements inside a styled container.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Radio from '@salo/core-ui/Forms/Radio';
```

Implement as follows

```javascript
<Radio
  name='checkbox_group_name'
  options={ [{ value:'val', label: 'Label' }] }
  onChange={ (value) => this.setState({value}) }
  value={ this.state.value }
>
```

## Asking for user input

In some cases it's useful to have an 'other' field where a user can type in whatever they like, this is accomplished like so:

```javascript
<Radio
  name='checkbox_group_name'
  options={ [{ value:'val', label: 'Label', value: 'other', label: 'Other', input: true }] }
  onChange={ (value) => this.setState({value}) }
  value={ this.state.value }
>
```

The key part is `input: true`, when this is detected and the option is chosen then the input will display.