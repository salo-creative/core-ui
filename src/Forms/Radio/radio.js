import React from 'react';
import PropTypes from 'prop-types';

// COMPONENTS & STYLES
import CheckBox from '../CheckBox';
import SaloInput from '../Input';
import ErrorText from '../components/ErrorText';
import HelperText from '../components/HelperText';
import Label from '../components/Label';
import { Group, GroupWrapper } from './radio.styles';

const Radio = props => {
  const {
    className,
    disabled,
    options,
    error,
    errorMessage,
    helperText,
    Input,
    label,
    name,
    margin,
    onChange,
    required,
    size,
    value
  } = props;

  const [other, setOther] = React.useState('');
  const [checked, setChecked] = React.useState(value);
  const inputEl = React.useRef();

  React.useEffect(() => {
    if (inputEl.current) {
      inputEl.current.focus();
    }
  });

  return (
    <Group
      className={ `${ className } salo-radio` }
      margin={ margin }
    >
      <Label
        error={ error }
        label={ label }
        name={ name }
        required={ required }
        size={ size }
        className='salo-radio__label'
      />
      <GroupWrapper className='salo-radio__wrapper'>
        { options.map((field, i) => {
          // Set margin equivalent to parent but hide for last item
          const checkMargin = i <= (options.length - 2) ? margin : '0';
          return (
            <React.Fragment key={ field.value }>
              <CheckBox
                checked={ field.value === checked }
                disabled={ disabled }
                label={ field.label }
                margin={ checkMargin }
                name={ field.value }
                onChange={ ({ name: val }) => {
                  onChange(val);
                  setChecked(val);
                  setOther('');
                } }
                radio
                size={ size }
              />
              { field.value === checked && field.input && (
                <Input
                  className='salo-radio__input'
                  margin='1rem 0 0'
                  ref={ inputEl }
                  name={ `${ name }-input` }
                  onBlur={ () => {
                    onChange(other);
                  } }
                  onChange={ ({ value: val }) => {
                    setOther(val);
                  } }
                  value={ other }
                />
              ) }
            </React.Fragment>
          );
        }) }
      </GroupWrapper>
      <ErrorText
        className='salo-radio__error'
        disabled={ disabled }
        error={ error }
        errorMessage={ errorMessage }
        size={ size }
      />
      <HelperText
        className='salo-radio__helper'
        disabled={ disabled }
        error={ error }
        helperText={ helperText }
        size={ size }
      />
    </Group>
  );
};

Radio.defaultProps = {
  className: '',
  disabled: false,
  error: false,
  errorMessage: 'Field invalid',
  helperText: '',
  Input: SaloInput,
  label: 'Label Text',
  margin: '0 0 2rem 0',
  onChange: () => null,
  options: [],
  required: false,
  size: 'M',
  value: ''
};

Radio.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  helperText: PropTypes.string,
  Input: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  label: PropTypes.string,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired
  })),
  required: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M']),
  value: PropTypes.string
};

export default Radio;