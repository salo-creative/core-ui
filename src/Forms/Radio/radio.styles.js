import styled from 'styled-components';
import { transparentize } from 'polished';

// Container
export const Wrapper = styled.div`
  margin: ${ ({ margin }) => margin };
`;

export const Group = styled.div`
  width: 100%;
  flex-direction: column;
  display: flex;
  margin: ${ ({ margin }) => margin };
`;

export const GroupWrapper = styled.div`
  padding: 1rem;
  background: ${ ({ theme }) => theme.paleGrey };
  border: 0.1rem solid;
  border-color: ${ ({ theme }) => transparentize(0.5, theme.primary) };
  border-radius: 0.4rem;
`;