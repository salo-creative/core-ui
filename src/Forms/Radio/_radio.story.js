import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, select } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// Tests.
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// Featured Components.
import { Radio } from '../../index';

// README
import README from './README.md';

// Story logic.
const stories = storiesOf('Forms | Radio', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['radio'] });

stories.add(
  'Basic',
  (() => {
    const options = [
      { value: 'field_1', label: 'Field 1' },
      { value: 'field_2', label: 'Field 2' },
      { value: 'field_3', label: 'Field 3' }
    ];
    // Store.
    const store = new Store({ value: '' });
    // Knobs.
    const label = text('Label', 'Radio group label');
    const helperText = text('Helper text', 'Helper text');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const size = select('size', ['L', 'M'], 'M');
    return (
      <RenderWithProps store={ store }>
        <Radio
          disabled={ disabled }
          error={ error }
          helperText={ helperText }
          label={ label }
          name='story'
          options={ options }
          onChange={ value => store.set({ value }) }
          required={ required }
          size={ size }
          value={ store.value }
        />
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);