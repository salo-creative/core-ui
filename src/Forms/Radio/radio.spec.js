import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Radio from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Radio />
  );
  expect(container.firstChild).toMatchSnapshot();
});