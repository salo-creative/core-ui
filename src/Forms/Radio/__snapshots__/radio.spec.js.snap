// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c1 {
  font-size: 1.4rem;
  margin: 0 0 0.5rem;
  padding: 0;
  color: inherit;
  word-break: break-word;
}

.c0 {
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 0 2rem 0;
}

.c2 {
  padding: 1rem;
  background: #fafcff;
  border: 0.1rem solid;
  border-color: rgba(0,172,237,0.5);
  border-radius: 0.4rem;
}

<div
  class="c0  salo-radio"
>
  <label
    class="c1 salo-radio__label"
  >
    Label Text
  </label>
  <div
    class="c2 salo-radio__wrapper"
  />
</div>
`;
